File: packages/platools-js/src/doctor/reporter.ts
/**
 * Text formatter for `platools doctor` reports.
 *
 * Matches the Python reporter output shape 1:1 so CI logs look the
 * same whichever SDK a consumer is on:
 *
 *     Tools: 47 registered, 43 healthy
 *
 *     ERRORS (2):
 *       x archive_workspace.admin_id is unreachable
 *       x merge_pages — circular dependency with split_page
 *
 *     WARNINGS (5):
 *       ! get_page — description too short (3 chars)
 *
 *     INFO (2):
 *       i cleanup_temp is not assigned to any agent
 */

import type { DoctorReport, Finding } from "./types.js";

export function formatReport(report: DoctorReport): string {
  const errors = report.errors();
  const warnings = report.warnings();
  const infos = report.infos();

  // Healthy count: tool count minus the unique tool names mentioned
  // in error-severity findings. Matches the Python reporter so
  // cycle-dedupe attributes unhealthy correctly (PLATOS-18
  // regression fix).
  const unhealthy = new Set<string>();
  for (const f of errors) {
    if (f.tool !== undefined) unhealthy.add(f.tool);
  }
  const healthy = report.toolCount - unhealthy.size;

  const lines: string[] = [];
  lines.push(`Tools: ${report.toolCount} registered, ${healthy} healthy`);
  lines.push("");

  if (errors.length > 0) {
    lines.push(`ERRORS (${errors.length}):`);
    lines.push(...formatSection(errors, "x"));
    lines.push("");
  }
  if (warnings.length > 0) {
    lines.push(`WARNINGS (${warnings.length}):`);
    lines.push(...formatSection(warnings, "!"));
    lines.push("");
  }
  if (infos.length > 0) {
    lines.push(`INFO (${infos.length}):`);
    lines.push(...formatSection(infos, "i"));
    lines.push("");
  }

  if (errors.length === 0 && warnings.length === 0 && infos.length === 0) {
    lines.push("No issues found.");
  }

  return `${lines.join("\n").replace(/\s+$/, "")}\n`;
}

function formatSection(findings: readonly Finding[], marker: string): string[] {
  return findings.map((f) => `  ${marker} ${f.message}`);
}

export function reportToJson(report: DoctorReport): string {
  return JSON.stringify(
    {
      tool_count: report.toolCount,
      errors: report.errors().map(findingJson),
      warnings: report.warnings().map(findingJson),
      info: report.infos().map(findingJson),
    },
    null,
    2,
  );
}

function findingJson(f: Finding): Record<string, unknown> {
  return {
    code: f.code,
    message: f.message,
    tool: f.tool ?? null,
    param: f.param ?? null,
  };
}

File: packages/platools-py/platools/doctor/reporter.py
"""Text formatter for `platools doctor` reports — PLATOS-18.

Matches the spec's CLI output shape:

    Tools: 47 registered, 43 healthy

    ERRORS (2):
      x archive_workspace.admin_id is unreachable
      x merge_pages — circular dependency with split_page

    WARNINGS (5):
      ! get_page — description too short (3 chars)

    INFO (2):
      i cleanup_temp is not assigned to any agent

The formatter is pure-text on purpose so the CLI can pipe it into
files / CI logs without ANSI noise. A `--color` flag in the CLI can
layer Rich on top later (PLATOS-32 follow-up); the data shape stays
the same.
"""

from __future__ import annotations

from platools.doctor.types import DoctorReport, Finding


def format_report(report: DoctorReport) -> str:
    """Render a `DoctorReport` as the canonical CLI text block."""
    errors = report.errors()
    warnings = report.warnings()
    infos = report.infos()
    healthy = report.tool_count - len({f.tool for f in errors if f.tool})

    lines: list[str] = []
    lines.append(f"Tools: {report.tool_count} registered, {healthy} healthy")
    lines.append("")

    if errors:
        lines.append(f"ERRORS ({len(errors)}):")
        lines.extend(_format_section(errors, marker="x"))
        lines.append("")
    if warnings:
        lines.append(f"WARNINGS ({len(warnings)}):")
        lines.extend(_format_section(warnings, marker="!"))
        lines.append("")
    if infos:
        lines.append(f"INFO ({len(infos)}):")
        lines.extend(_format_section(infos, marker="i"))
        lines.append("")

    if not (errors or warnings or infos):
        lines.append("No issues found. 🎉")

    return "\n".join(lines).rstrip() + "\n"


def _format_section(findings: list[Finding], *, marker: str) -> list[str]:
    return [f"  {marker} {finding.message}" for finding in findings]
